### PR TITLE
Allow instance-id to auto-generate typings, separate internal vs external APIs

### PR DESCRIPTION
**Pull request by MathBunny**
## Context
The goal is to bridge the gap between today's Admin SDK (manually curated d.ts files, nested namespaces) and a modularized SDK that has auto-generated typings.

This is the second milestone from: `go/firebase-node-auto-typing`.

## Goals for this PR
1. Allow to auto-generated typings for instance-id
2. Split the internal and external APIs
3. Still use the previous manually curated typings
4. Be a non-breaking change

## Prerequisites
~The intention for this PR is to be merged after RTDB modularization.~
The gulpfile PR from `hlazu-modularizing-gulpfile`.

## Test Plan
1. Unit tests: 
> `npm test`
2. Integration tests: 
> `export TYPE_GENERATION_MODE=nonauto && npm run integration && export TYPE_GENERATION_MODE=auto && npm run integration`
3. Sanity tests:  
> `export TYPE_GENERATION_MODE=nonauto && npm pack && ./.github/scripts/verify_package.sh firebase-* && export TYPE_GENERATION_MODE=auto && npm pack && ./.github/scripts/verify_package.sh firebase-*`

### CODE REVIEW

**Comment by hiranya911** [comment]
Why did this require changes to Auth typings? That shouldn't be related to `instanceId` module.

**Comment by MathBunny** [comment]
Sorry, this PR isn't ready for review yet (still in draft), but I mostly came across them while working on this because the type-checker brought it up when I tried to interchange `admin.app.App` and `FirebaseApp`, the changes won't be included in the final PR. I created a separate PR for the auth typings.

File: src/instance-id/instance-id-request.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*!
- * Copyright 2017 Google Inc.
+ * Copyright 2020 Google Inc.
```
Comment (hiranya911) [inline src/instance-id/instance-id-request.ts:2]: Unrelated to this change. Let's revert this.

File: src/instance-id/instance-id-internal.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+   *
+   * @return {Promise<()>} An empty Promise that will be fulfilled when the service is deleted.
+   */
+  public delete(): Promise<void> {
+    // There are no resources to clean up
+    return Promise.resolve(undefined);
+  }
+}
+
+export class InstanceId implements FirebaseServiceInterface {
```
Comment (hiranya911) [inline src/instance-id/instance-id-internal.ts:39]: This should be called `InstanceIdImpl` and also implement the `InstanceId` interface.

File: src/firebase-namespace.ts [hunk 3]
```diff
@@ -33,7 +33,7 @@ import { Messaging } from './messaging/messaging';
 import { Storage } from './storage/storage';
 import { Database } from '@firebase/database';
 import { Firestore } from '@google-cloud/firestore';
-import { InstanceId } from './instance-id/instance-id';
+import { InstanceId } from './instance-id/instance-id-internal';
```
Comment (hiranya911) [inline src/firebase-namespace.ts:36]: This should be the interface type.

File: src/firebase-app.ts [hunk 4]
```diff
@@ -29,7 +29,7 @@ import { Database } from '@firebase/database';
 import { DatabaseService } from './database/database';
 import { Firestore } from '@google-cloud/firestore';
 import { FirestoreService } from './firestore/firestore';
-import { InstanceId } from './instance-id/instance-id';
+import { InstanceId }  from './instance-id/instance-id-internal';
```
Comment (hiranya911) [inline src/firebase-app.ts:32]: This should be the interface type.

File: src/firebase-app.ts [hunk 5]
```diff
@@ -352,7 +352,7 @@ export class FirebaseApp {
    */
   public instanceId(): InstanceId {
     return this.ensureService_('iid', () => {
-      const iidService: typeof InstanceId = require('./instance-id/instance-id').InstanceId;
+      const iidService: typeof InstanceId = require('./instance-id/instance-id-internal').InstanceId;
```
Comment (hiranya911) [inline src/firebase-app.ts:355]: This should look something like:

```
import { InstanceId } from './instance-id/instance-id';
import { InstanceIdImpl } from './instance-id/instance-id-internal';

const iidService: typeof InstanceIdImpl = require('./instance-id/instance-id-internal').InstanceIdImpl;
return new iidService(this);
```

File: test/unit/firebase-app.spec.ts [hunk 6]
```diff
@@ -36,7 +36,7 @@ import { MachineLearning } from '../../src/machine-learning/machine-learning';
 import { Storage } from '../../src/storage/storage';
 import { Firestore } from '@google-cloud/firestore';
 import { Database } from '@firebase/database';
-import { InstanceId } from '../../src/instance-id/instance-id';
+import { InstanceId } from '../../src/instance-id/instance-id-internal';
```
Comment (hiranya911) [inline test/unit/firebase-app.spec.ts:39]: Should be the interface type.

File: test/unit/firebase-namespace.spec.ts [hunk 7]
```diff
@@ -48,7 +48,7 @@ import {
   v1beta1,
   setLogFunction,
 } from '@google-cloud/firestore';
-import { InstanceId } from '../../src/instance-id/instance-id';
+import { InstanceId } from '../../src/instance-id/instance-id-internal';
```
Comment (hiranya911) [inline test/unit/firebase-namespace.spec.ts:51]: Here too.

File: src/instance-id/instance-id-internal.ts (hunk 2 above)
Comment (MathBunny) [reply src/instance-id/instance-id-internal.ts]: That's correct, I completely missed this part.

File: test/unit/firebase-namespace.spec.ts (hunk 7 above)
Comment (MathBunny) [reply test/unit/firebase-namespace.spec.ts]: Fixed, though I ended up using both the interface and implementation for this:

https://github.com/firebase/firebase-admin-node/blob/560f1061669b6172f5f07c656d5934cae4f65c9a/test/unit/firebase-namespace.spec.ts#L626-L628

**Comment by MathBunny** [comment]
Thanks for the feedback, PTAL :)

**Comment by MathBunny** [comment]
Rebased, PTAL @hiranya911  :)

**Pull request merged**